docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
- return &KVStoreApplication{
-  db: db,
- }
+And call it from within your `CheckTx` method:
+
+```go
+func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
+	if code := app.validateTx(req.Tx); code != 0 {
+		return abcitypes.ResponseCheckTx{Code: code}
+	}
+	return abcitypes.ResponseCheckTx{Code: 0}
 }
 ```
 
+Any response with a non-zero code will be considered invalid by Tendermint. 
+Our `CheckTx` logic returns `0` to Tendermint when the transaction passes
+the validation checks.
+
 ### 1.3.2 BeginBlock -> DeliverTx -> EndBlock -> Commit
 
-When Tendermint Core has decided on the block, it's transfered to the
-application in 3 parts: `BeginBlock`, one `DeliverTx` per transaction and
-`EndBlock` in the end. DeliverTx are being transfered asynchronously, but the
-responses are expected to come in order.
+When the Tendermint consensus engine has decided on the block, the block is transferred to the
+application over three ABCI method calls: `BeginBlock`, `DeliverTx`, and `EndBlock`.
+
+`BeginBlock` is called once to indicate to the application that it is about to
+receive a block.
+
+`DeliverTx` is called repeatedly, once for each `Tx` that was included in the block.
+
+`EndBlock` is called once to indicate to the application that no more transactions
+will be delivered to the application.
+
+To implement these calls in our application we're going to make use of Badger's 
+transaction mechanism. 
+
+First, let's create a new Badger `Txn` during `BeginBlock`:
 
 ```go
 func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
- app.currentBatch = app.db.NewTransaction(true)
- return abcitypes.ResponseBeginBlock{}
+	app.currentBatch = app.db.NewTransaction(true)
+	return abcitypes.ResponseBeginBlock{}
 }
-
 ```
 
-Here we create a batch, which will store block's transactions.
+Next, let's modify `DeliverTx` to add the `key` and `value` to this `Txn` every time our application
+receives a new `RequestDeliverTx`.
 
 ```go
 func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
- code := app.isValid(req.Tx)
- if code != 0 {
-  return abcitypes.ResponseDeliverTx{Code: code}
- }
+	if code := app.validateTx(req.Tx); code != 0 {
+		return abcitypes.ResponseDeliverTx{Code: code}
+	}
 
- parts := bytes.Split(req.Tx, []byte("="))
- key, value := parts[0], parts[1]
+	parts := bytes.Split(req.Tx, []byte("="))
+	key, value := parts[0], parts[1]
 
- err := app.currentBatch.Set(key, value)
- if err != nil {
-  panic(err)
- }
+	err := app.currentBatch.Set(key, value)
+	if err != nil {
+		panic(err)
+	}
 
- return abcitypes.ResponseDeliverTx{Code: 0}
+	return abcitypes.ResponseDeliverTx{Code: 0}
 }
 ```
+Note that we check the validity of the transaction _again_ during `DeliverTx`.
+Transactions are not guaranteed to be valid when they are delivered to to an
+application. 
```

Comment:
Is there anything we can say here about why that could be, even if only cursory?

I don't think we need to give deep details, but having a general sense of how a tx could be invalid at delivery even if it passed `CheckTx` might help build intuition.